### PR TITLE
Convert mode to typescript

### LINKED ISSUE #22912: Convert Mode to TypeScript
Convert [Mode](https://github.com/metabase/metabase/blob/master/frontend/src/metabase-lib/lib/Mode.ts) to TypeScript.

### PR DESCRIPTION
**Pull request by losrebellos**
Fixes #22912.

### CODE REVIEW

File: frontend/src/types/object-literal.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+type ObjectLiteral<Type = any> = {
```
Comment (losrebellos) [inline frontend/src/types/object-literal.d.ts:1]: Added typed object literal type

File: frontend/src/metabase-lib/lib/Mode/utils.ts [hunk 2]
```diff
@@ -1,35 +1,40 @@
...
+  MODE_TYPE_SEGMENT,
+  MODE_TYPE_METRIC,
+  MODE_TYPE_TIMESERIES,
+  MODE_TYPE_GEO,
+  MODE_TYPE_PIVOT,
+  MODE_TYPE_DEFAULT,
+} from "metabase-lib/lib/Mode/constants";
 
-export function getMode(question) {
+export function getMode(question: Question): ModeType | null {
```
Comment (losrebellos) [inline frontend/src/metabase-lib/lib/Mode/utils.ts:15]: Returns the mode type only, thus making it independent from `metabase-types/types/Visualization` and `frontend/src/metabase/modes`.

File: frontend/src/metabase/modes/lib/modes.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import {
+  MODE_TYPE_NATIVE,
+  MODE_TYPE_SEGMENT,
+  MODE_TYPE_METRIC,
+  MODE_TYPE_TIMESERIES,
+  MODE_TYPE_GEO,
+  MODE_TYPE_PIVOT,
+} from "metabase-lib/lib/Mode/constants";
+
+export function getMode(question: Question): QueryMode | any | null {
```
Comment (losrebellos) [inline frontend/src/metabase/modes/lib/modes.ts:21]: Returns `QueryMode | any | null` but in effect should only returns `QueryMode | null`.
All modes files are not TS and does not support typing atm.

**Review by npfitz — APPROVED** [review]

File: frontend/src/metabase/modes/lib/modes.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import SegmentMode from "../components/modes/SegmentMode";
+import MetricMode from "../components/modes/MetricMode";
+import TimeseriesMode from "../components/modes/TimeseriesMode";
+import GeoMode from "../components/modes/GeoMode";
+import PivotMode from "../components/modes/PivotMode";
+import NativeMode from "../components/modes/NativeMode";
+import DefaultMode from "../components/modes/DefaultMode";
+import { QueryMode } from "metabase-types/types/Visualization";
```
Comment (ranquild) [inline frontend/src/metabase/modes/lib/modes.ts:8]: Deprecated `metabase-types/types` usage. Please use types from `metabase-types/api`. If they don't exist yet, please move this type there.

File: frontend/src/types/object-literal.d.ts (hunk 1 above)
Comment (ranquild) [inline frontend/src/types/object-literal.d.ts:1]: There is a built-in `Record` type, can it be used here instead?

**Review by ranquild — CHANGES_REQUESTED** [review]
Usage of deprecated `metabase-types/types` and a question about `Record`

File: frontend/src/metabase/modes/lib/modes.ts (hunk 4 above)
Comment (losrebellos) [reply frontend/src/metabase/modes/lib/modes.ts:8]: I had a look at moving the necessary `types` to `api` but it seems to be a very big job as there is a lots of dependencies and out of the scope regarding that issue.

**Review by ranquild — APPROVED** [review]

**Review by daltojohnso — APPROVED** [review]

**Pull request merged**